applications/services/gui/modules/number_input.c

```diff
@@ -0,0 +1,493 @@
+    if(model->selected_row > 0) {
+        model->selected_row--;
+        if(model->selected_column > number_input_get_row_size(model->selected_row) - 1) {
+            model->selected_column = number_input_get_row_size(model->selected_row) - 1;
+        }
+    }
+}
+
+/** Handle down button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_down(NumberInputModel* model) {
+    if(model->selected_row < keyboard_row_count - 1) {
+        if(model->selected_column >= number_input_get_row_size(model->selected_row) - 1) {
+            model->selected_column = number_input_get_row_size(model->selected_row + 1) - 1;
+        }
+        model->selected_row += 1;
+    }
+    const NumberInputKey* keys = number_input_get_row(model->selected_row);
+    if(keys[model->selected_column].text == sign_symbol && !model->useSign) {
+        model->selected_column--;
+    }
+}
+
+/** Handle left button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_left(NumberInputModel* model) {
+    if(model->selected_column > 0) {
+        model->selected_column--;
+    } else {
+        model->selected_column = number_input_get_row_size(model->selected_row) - 1;
+    }
+    const NumberInputKey* keys = number_input_get_row(model->selected_row);
+    if(keys[model->selected_column].text == sign_symbol && !model->useSign) {
+        model->selected_column--;
+    }
+}
+
+/** Handle right button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_right(NumberInputModel* model) {
+    if(model->selected_column < number_input_get_row_size(model->selected_row) - 1) {
+        model->selected_column++;
+    } else {
+        model->selected_column = 0;
+    }
+    const NumberInputKey* keys = number_input_get_row(model->selected_row);
+    if(keys[model->selected_column].text == sign_symbol && !model->useSign) {
+        model->selected_column++;
+    }
+}
+
+static void prevent_to_large_number(NumberInputModel* model) {
+    if(strtol(furi_string_get_cstr(model->text_buffer), NULL, 10) > model->max_value) {
+        char* str = int32_to_string(model->max_value);
+        furi_string_set_str(model->text_buffer, str);
+        free(str);
+    }
+    // Added in prevent large, as it would block the input of small positive numbers
+    if(model->sign == '-' &&
+       strtol(furi_string_get_cstr(model->text_buffer), NULL, 10) < model->min_value) {
+        char* str = int32_to_string(model->min_value);
+        furi_string_set_str(model->text_buffer, str);
+        free(str);
+    }
+}
+
+static void prevent_to_small_number(NumberInputModel* model) {
+    if(strtol(furi_string_get_cstr(model->text_buffer), NULL, 10) < model->min_value) {
+        char* str = int32_to_string(model->min_value);
+        furi_string_set_str(model->text_buffer, str);
+        free(str);
+    }
+}
+
+static void number_input_sign(NumberInputModel* model) {
+    if(model->sign == '-') {
+        model->sign = '+';
```

Comment:
There is a model and there is a representation, what you doing is mixing representation into model.